server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala

```diff
@@ -39,12 +44,50 @@ case class PlainPart(name: String) extends JsonPathPart {
   }
 }
 
+object ArrayElementPart {
+  def parse(string: String): Option[ArrayElementPart] = {
+    if (string.startsWith("[") && string.endsWith("]")) {
+      val positionPart: String = string.substring(1, string.length - 1)
+      Try(positionPart.toInt)
+        .fold(_ => None, fromInt)
+    } else {
+      None
+    }
+  }
+
+  private def fromInt(position: Int): Option[ArrayElementPart] =
+    refineV[NonNegative](position)
+      .fold(_ => None,
+        ref => Some(ArrayElementPart(ref)))
+}
+
+case class ArrayElementPart(position: NonNegInt) extends JsonPathPart {
+  def read(jsValue: JsValue): JsResult[JsValue] = jsValue match {
+    case JsArray(values) => values.lift(position.value)
+      .map(JsSuccess(_))
+      .getOrElse(JsError(s"Supplied array have no $position element"))
+    case _ => JsError("Expecting a JsArray but got a different structure")
+  }
+}
+
 object JsonPath {
   def parse(string: String): JsonPath = JsonPath(string.split('/').toList
     .flatMap {
-      case "" => None
-      case "*" => Some(WildcardPart())
-      case part => Some(PlainPart(part))
+      case "" => Nil
+      case "*" => List(WildcardPart())
+      case string if ArrayElementPart.parse(string).isDefined => ArrayElementPart.parse(string)
+      case part: String =>
+        val arrayElementPartPosition = part.indexOf('[')
+        if (arrayElementPartPosition < 0) {
+          List(PlainPart(part))
+        } else if (arrayElementPartPosition == 0) {
+          List(ArrayElementPart.parse(string)
+            .getOrElse(PlainPart(string)))
+        } else {
+          ArrayElementPart.parse(string.substring(arrayElementPartPosition))
```

Comment:
Each case could be extracted in a method to put a name on them. Which would render this method more readable